apps/application-system/api/src/app/modules/application/application.controller.ts

```diff
@@ -404,86 +426,203 @@ export class ApplicationController {
       answers: mergedAnswers,
     }
 
-    const [hasChanged, updatedApplication] = await this.changeState(
+    const {
+      hasChanged,
+      hasError,
+      error,
+      application: updatedApplication,
+    } = await this.changeState(
       mergedApplication,
       template,
       updateApplicationStateDto.event,
       authorization,
     )
 
-    // TODO: should not have to specificially check for updatedApplication
-    // because of return type on this.changeState
-    if (hasChanged === true && updatedApplication) {
+    if (hasError) {
+      throw new BadRequestException(error)
+    }
+
+    if (hasChanged) {
       return updatedApplication
     }
 
     return existingApplication
   }
 
-  async changeState(
+  async performActionOnApplication(
     application: BaseApplication,
     template: Unwrap<typeof getApplicationTemplateByTypeId>,
-    event: string,
     authorization: string,
-  ): Promise<[false] | [true, BaseApplication]> {
-    const helper = new ApplicationTemplateHelper(application, template)
+    action: ApplicationTemplateAPIAction,
+  ): Promise<TemplateAPIModuleActionResult> {
+    const {
+      apiModuleAction,
+      shouldPersistToExternalData,
+      externalDataId,
+      throwOnError,
+    } = action
+
+    const actionResult = await this.templateAPIService.performAction({
+      templateId: template.type,
+      type: apiModuleAction,
+      props: {
+        application,
+        authorization,
+      },
+    })
 
-    const [hasChanged, newState, newApplication] = helper.changeState(event)
+    let updatedApplication: BaseApplication = application
 
-    if (!hasChanged) {
-      return [false]
+    if (shouldPersistToExternalData) {
+      const newExternalDataEntry: ExternalData = {
+        [externalDataId || apiModuleAction]: {
+          status: actionResult.success ? 'success' : 'failure',
+          date: new Date(),
+          data: actionResult.success
+            ? (actionResult.response as ExternalData['data'])
+            : actionResult.error,
+        },
+      }
+
+      const {
+        updatedApplication: withExternalData,
+      } = await this.applicationService.updateExternalData(
+        updatedApplication.id,
+        updatedApplication.externalData,
+        newExternalDataEntry,
+      )
+
+      updatedApplication = {
+        ...updatedApplication,
+        externalData: {
+          ...updatedApplication.externalData,
+          ...withExternalData.externalData,
```

Comment:
I might be wrong on this, but there is no way the API action is returning an object it shouldn't handle and overwrite it by mistake? if API module should return objA and is returning objA and objB, won't objB overwritten?

Reply:
This happens after updating the application's external data in the database and is only updating the `updatedApplication` externalData with whatever it confirmed to be updated in the database.

There are two cases where unrelated external data would get updated:
- The developer named the action the same as the id of an existing DataProvider
- The developer chose a custom `externalDataId` with the same id as an existing DataProvider